### PR TITLE
#1227 Explicit Logic for Communication Items

### PR DESCRIPTION
**Pull request by EvanParish**
# Description

This PR is to add logic that will change when notifications are sent depending on the "default send" preference set in the communication item attached to the template. 

#1227

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

Deployed to dev and notifications sent with below parameters.
User’s VA Profile ID 43627 has no selection for any communication items
- [x] TC01 EMAIL: template “Email Default Send True”; expect a successful send and CommunicationItemNotFoundException is seen in logs.
- [x] TC02 EMAIL: template “Email Default Send False”; expect an unsuccessful send, the status is set to "preference-declined" with status "No recipient opt-in found for explicit preference", and CommunicationItemNotFoundException is seen in logs.
- [x] TC03 SMS: template “SMS Default Send True”; expect a successful send and CommunicationItemNotFoundException is seen in logs.
- [x] TC04 SMS: template “SMS Default Send False”; expect an unsuccessful send, the status is set to "preference-declined" with status "No recipient opt-in found for explicit preference", and CommunicationItemNotFoundException is seen in logs.
- [x] TC13 EMAIL: template "Email with custom reply-to"; expect a successful send and CommunicationItemNotFoundException is seen in logs.

User’s VA Profile 112491 has the above communication items set to “allowed”: “true”
- [x] TC05 EMAIL: template “Email Default Send True”; expect a successful send.
- [x] TC06 EMAIL: template “Email Default Send False”; expect a successful send.
- [x] TC07 SMS: template “SMS Default Send True”; expect a successful send.
- [x] TC08 SMS: template “SMS Default Send False”; expect a successful send.

User’s VA Profile 37459 has the has the above communication items set to “allowed”: “false”
- [x] TC09 EMAIL: template “Email Default Send True”; expect an unsuccessful send, with status preference-declined.
- [x] TC11 EMAIL: template “Email Default Send False”; expect an unsuccessful send, with status preference-declined.
- [x] TC10 SMS: template “SMS Default Send True”; expect an unsuccessful send, with status preference-declined.
- [x] TC12 SMS: template “SMS Default Send False”; expect an unsuccessful send, with status preference-declined.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by cris-oddball** [comment]
@EvanParish you'll need to rebase or something to clear that dependency-scan error, it's for flask and we merged that into master yesterday.

File: tests/app/celery/test_lookup_recipient_communication_permissions_task.py [hunk 1]
```diff
@@ -126,9 +128,69 @@ def test_recipient_has_given_permission_should_return_true_if_user_grants_permis
...
+):
+    mocked_va_profile_client = mocker.Mock(VAProfileClient)
+    mocked_va_profile_client.get_is_communication_allowed = mocker.Mock(side_effect=CommunicationItemNotFoundException)
+    mocker.patch(
+        'app.celery.lookup_recipient_communication_permissions_task.va_profile_client',
+        new=mocked_va_profile_client
+    )
+
+    mock_communication_item = mocker.Mock(CommunicationItem)
+    mock_communication_item.default_send_indicator = send_indicator
```
Comment (kalbfled) [inline tests/app/celery/test_lookup_recipient_communication_permissions_task.py:178]: This doesn't need to be mocked at all.  You could just create a new CommunicationItem instance.

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 2]
```diff
@@ -58,18 +61,37 @@ def recipient_has_given_permission(
         notification_id: str,
         notification_type: str,
         communication_item_id: str
-) -> bool:
+) -> str:
```
Comment (kalbfled) [inline app/celery/lookup_recipient_communication_permissions_task.py:64]: Should be `Optional[str]`.

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 3]
```diff
@@ -58,18 +61,37 @@ def recipient_has_given_permission(
...
-    communication_item = get_communication_item(communication_item_id)
+    try:
+        communication_item = get_communication_item(communication_item_id)
+    except NoResultFound:
+        current_app.logger.info('No communication item found for notification %s', notification_id)
 
     try:
+        # get default send flag when available
+        if communication_item is not None:
+            default_send_flag = communication_item.default_send_indicator
```
Comment (kalbfled) [inline app/celery/lookup_recipient_communication_permissions_task.py:77]: Only code that might raise an exception should be in a "try" block.

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 4]
```diff
@@ -58,18 +61,37 @@ def recipient_has_given_permission(
...
+
         is_allowed = va_profile_client.get_is_communication_allowed(
```
Comment (kalbfled) [inline app/celery/lookup_recipient_communication_permissions_task.py:79]: If communication_item is None, does this always raise an exception?  If so, only call it if `communication_item is not None`.

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 5]
```diff
@@ -58,18 +61,37 @@ def recipient_has_given_permission(
...
+            notification_id,
+            notification_type
         )
-        current_app.logger.info(f'Value of permission for item {communication_item.va_profile_item_id} for recipient '
-                                f'{id_value} for notification {notification_id}: {is_allowed}')
-        return is_allowed
+
+        current_app.logger.info('Value of permission for item %s for recipient %s for notification %s: %s',
+                                communication_item.va_profile_item_id if communication_item else None,
+                                id_value, notification_id, is_allowed)
```
Comment (kalbfled) [inline app/celery/lookup_recipient_communication_permissions_task.py:88]: I would move this to the "else" block of a try-except-else structure.

**Review by kalbfled — CHANGES_REQUESTED** [review]

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 6]
```diff
@@ -116,3 +111,13 @@ def recipient_has_given_permission(
...
+    else:
+        current_app.logger.info('Value of permission for item %s for recipient %s for notification %s: %s',
+                                communication_item.va_profile_item_id if communication_item else None,
+                                id_value, notification_id, is_allowed)
+
+        if is_allowed:
+            return None
+
+        # return status reason message if message should not be sent
+        return "Contact preferences set to false"
```
Comment (kalbfled) [inline app/celery/lookup_recipient_communication_permissions_task.py:123]: As revised, now you don't need an else block because either "except" block will exit the code.  You only get here if no exception occurred.  Also note that you no longer need to test `if communication_item` because you already returned above if that value was None.

Finally, I would consolidate the return value:
```
return None if is_allowed else "Contact preferences set to false"
```

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 7]
```diff
@@ -71,27 +72,21 @@ def recipient_has_given_permission(
...
-        # get default send flag when available
-        if communication_item is not None:
-            default_send_flag = communication_item.default_send_indicator
+    # get default send flag when available
+    if communication_item is not None:
+        default_send_flag = communication_item.default_send_indicator
+    else:
+        # Calling va_profile without a communication item won't return anything.
+        # Perform default behavior of sending the notification.
+        return None
```
Comment (kalbfled) [inline app/celery/lookup_recipient_communication_permissions_task.py:81]: Slightly more concise:
```
if communication_item is None:
    # Your comments
    return None

default_send_flag = ...
```

File: tests/app/celery/test_lookup_recipient_communication_permissions_task.py [hunk 8]
```diff
@@ -174,12 +174,12 @@ def test_recipient_has_given_permission_with_default_send_indicator_and_no_prefe
         new=mocked_va_profile_client
     )
 
-    mock_communication_item = mocker.Mock(CommunicationItem)
-    mock_communication_item.default_send_indicator = send_indicator
+    test_communication_item = CommunicationItem
+    test_communication_item.default_send_indicator = send_indicator
```
Comment (kalbfled) [inline tests/app/celery/test_lookup_recipient_communication_permissions_task.py:178]: This doesn't seem to fail the test, but it's passing by accident.  "CommunicationItem" is the class, not an instance, and you've made "default_send_indicator" a static attribute.  You should instead instantiate:
```
test_communication_item = CommunicationItem(id=uuid4(), va_profile_item_id=1, name="name", default_send_indicator=send_indicator)
```

```
>>> class Test:
...     pass
... 
>>> Test.static_attr = "hello"

>>> a = Test()
>>> b = Test()

>>> a.static_attr
'hello'
>>> b.static_attr
'hello'

>>> b.static_attr = "change this"
>>> a.static_attr
'hello'
>>> b.static_attr
'change this'

>>> Test.static_attr = "something else"
>>> a.static_attr
'something else'
>>> b.static_attr
'change this'
```

**Review by kalbfled — CHANGES_REQUESTED** [review]

File: tests/app/celery/test_lookup_recipient_communication_permissions_task.py [hunk 9]
```diff
@@ -110,9 +111,10 @@ def test_recipient_has_given_permission_should_return_false_if_user_denies_permi
     )
 
     mock_task = mocker.Mock()
-    assert not recipient_has_given_permission(
+    permission_message = recipient_has_given_permission(
```
[the remark this answers is not in the document]
Comment (EvanParish) [reply tests/app/celery/test_lookup_recipient_communication_permissions_task.py:114]: updated

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 10]
```diff
@@ -72,21 +72,25 @@ def recipient_has_given_permission(
...
         is_allowed = va_profile_client.get_is_communication_allowed(
-            identifier,
-            communication_item.va_profile_item_id if communication_item is not None else None,
-            notification_id,
-            notification_type
+            identifier, communication_item.va_profile_item_id, notification_id, notification_type
         )
+
+        current_app.logger.info('Value of permission for item %s for recipient %s for notification %s: %s',
+                                communication_item.va_profile_item_id if communication_item else None,
```
Comment (kalbfled) [inline app/celery/lookup_recipient_communication_permissions_task.py:89]: You don't need to test "communication_item".  You already know it isn't None because you tested above.

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 11]
```diff
@@ -106,18 +110,5 @@ def recipient_has_given_permission(
...
 
-        if default_send_flag:
-            return None
-
-        # return status reason message if message should not be sent
-        return "No recipient opt-in found for explicit preference"
-    else:
-        current_app.logger.info('Value of permission for item %s for recipient %s for notification %s: %s',
-                                communication_item.va_profile_item_id if communication_item else None,
-                                id_value, notification_id, is_allowed)
```
Comment (kalbfled) [inline app/celery/lookup_recipient_communication_permissions_task.py]: You moved this back inside the "try".  My apologies if this wasn't clear, but it still belongs here.  I intended to suggest that you delete the "else" and de-indent here.

**Review by kalbfled — CHANGES_REQUESTED** [review]

**Review by kalbfled — DISMISSED** [review]

**Review by kalbfled — APPROVED** [review]

**Review by cris-oddball — APPROVED** [review]
